melodies_monet/util/sat_l2_swath_utility_tempo.py

```diff
@@ -0,0 +1,899 @@
+
+    scattering_weights = obsobj["scattering_weights"].transpose("swt_level", "x", "y")
+    scattering_weights = scattering_weights.rename({"swt_level": "z"})
+    amf = obsobj["amf"]
+    modhcho_col = (scattering_weights * partial_col).sum(dim="z") / amf
+    modhcho_col = modhcho_col.where(modobj[f"{species}_col"].isel(z=0).notnull())
+    modhcho_col.attrs = {
+        "units": "molecules/cm2",
+        "description": "model HCHO column after applying TEMPO scattering weights and AMF",
+        "history": "Created by MELODIES-MONET, apply_weights_mod2tempo_hcho, TEMPO util",
+    }
+    return modhcho_col.where(np.isfinite(modhcho_col))
+
+
+def is_nonpairable(obsobj, k, modobj):
+    """Discards inplace granules from obsobj that do not match modobj's
+    domain, or granules that are all NaN. If the domain is small,
+    it can considerably speed up the regridding process.
+
+    Parameters
+    ----------
+    obsobj : dict[str, xr.Dataset]
+        tempo data
+    modobj : xr.Dataset
+        model data
+
+    Return
+    ------
+    collections.OrderedDict[str, xr.Dataset]
+    """
+    if obsobj[k]["lon"].max() < modobj["longitude"].min():
+        return True
+    elif obsobj[k]["lon"].min() > modobj["longitude"].max():
+        return True
+    elif obsobj[k]["lat"].max() < modobj["latitude"].min():
+        return True
+    elif obsobj[k]["lat"].min() > modobj["latitude"].max():
+        return True
+    return False
+
+
+def _regrid_and_apply_weights(
+    obsobj, modobj, method="conservative", weights=None, species=["NO2"], tempo_sp="NO2"
+):
+    """Does the complete process of regridding and
+    applying scattering weights. Assumes that obsobj is a Dataset
+
+    Parameters
+    ----------
+    obsobj : xr.Dataset
+        TEMPO observations
+    modobj : xr.Dataset
+        Model data
+    method : str
+        Choose regridding method. Can be "conservative", "conservative_normed",
+        "bilinear" or "patch". Check xesmf documentation for details.
+    weights : str
+        Path to the weightfile. If present, the weights won't be calculated again.
+    tempo_sp: str
+        NO2 or HCHO, to apply the correct Air Mass Factors and scattering weights.
+
+    Returns
+    -------
+    xr.DataArray
+        Model data regridded to the TEMPO grid,
+        with the averaging kernel.
+    """
+    if tempo_sp == "NO2":
+        apply_weights = apply_weights_mod2tempo_no2
+        apply_weights_hydrostatic = apply_weights_mod2tempo_no2_hydrostatic
+    else:
+        assert tempo_sp == "HCHO", "TEMPO species must be HCHO or NO2."
+        apply_weights = apply_weights_mod2tempo_hcho
+        apply_weights_hydrostatic = apply_weights_mod2tempo_hcho_hydrostatic
+    if method == "conservative":
+        if "lat_b" not in modobj:
+            modobj["lat_b"], modobj["lon_b"] = calc_grid_corners(modobj)
+        if "lat_b" not in obsobj:
+            obsobj["lat_b"], obsobj["lon_b"] = calc_grid_corners(obsobj, lat="lat", lon="lon")
```

Comment:
if change calc_grid_corners to work inplace, there isn't anything returned and the code needs to be adjusted here to reflect.
```suggestion
            calc_grid_corners(modobj)
        if "lat_b" not in obsobj:
            calc_grid_corners(obsobj, lat="lat", lon="lon")
```

Reply:
Looks good